core/src/main/java/org/infinispan/interceptors/distribution/L1NonTxInterceptor.java

```diff
@@ -314,25 +314,20 @@ private Object handleWriteManyCommand(InvocationContext ctx, WriteCommand comman
    @Override
    public Object visitInvalidateL1Command(InvocationContext ctx, InvalidateL1Command invalidateL1Command)
          throws Throwable {
-      AggregateCompletionStage<Void> aggregateCompletionStage = null;
+      CompletableFuture<Void> initialStage = new CompletableFuture<>();
+      CompletionStage<Void> currentStage = initialStage;
       for (Object key : invalidateL1Command.getKeys()) {
          abortL1UpdateOrWait(key);
          // If our invalidation was sent when the value wasn't yet cached but is still being requested the context
          // may not have the value - if so we need to add it then now that we know we waited for the get response
          // to complete
          if (ctx.lookupEntry(key) == null) {
-            CompletionStage<Void> stage = entryFactory.wrapEntryForWriting(ctx, key, keyPartitioner.getSegment(key),
-                  true, false);
-            if (!CompletionStages.isCompletedSuccessfully(stage)) {
-               if (aggregateCompletionStage == null) {
-                  aggregateCompletionStage = CompletionStages.aggregateCompletionStage();
-               }
-               aggregateCompletionStage.dependsOn(stage);
-            }
+            currentStage = entryFactory.wrapEntryForWriting(ctx, key, keyPartitioner.getSegment(key),
+                                                            true, false, currentStage);
          }
       }
-      return aggregateCompletionStage != null ? asyncInvokeNext(ctx, invalidateL1Command, aggregateCompletionStage.freeze()) :
-            invokeNext(ctx, invalidateL1Command);
+      if (currentStage == initialStage) return invokeNext(ctx, invalidateL1Command);
+      return asyncInvokeNext(ctx, invalidateL1Command, currentStage);
```

Comment:
isn't `expirationCheckDelay(currentStage, initialStage)` missing?

Reply:
Yep, when I wrote this `expirationCheckDelay()` was in `EntryWrappingInterceptor` so I "inlined" it here, but I didn't update it when I modified the original later and now it's missing the `initialStage.complete(null)` call.